pwiz_tools/Skyline/Controls/Graphs/ReplicateCachingReceiver.cs

```diff
@@ -0,0 +1,310 @@
+using System.Threading;
+using pwiz.Common.SystemUtil.Caching;
+using pwiz.Skyline.Model;
+using pwiz.Skyline.Util;
+
+namespace pwiz.Skyline.Controls.Graphs
+{
+    /// <summary>
+    /// Wraps a Receiver with a local cache to enable fast replicate switching.
+    ///
+    /// The standard Producer/Receiver pattern only caches the most recent result.
+    /// When switching between replicates (e.g., replicate 1 → 2 → 1), each switch
+    /// triggers a full recalculation because the old result is discarded.
+    ///
+    /// This wrapper maintains a local cache keyed by replicate index, allowing
+    /// instant switching between previously-viewed replicates. The cache is
+    /// invalidated when the document or settings change.
+    ///
+    /// Additionally, when switching away from a calculation in progress, this wrapper
+    /// keeps a "completion listener" attached so the calculation continues in the
+    /// background and its result is cached. This means switching away and back
+    /// doesn't lose progress - like browser tabs that keep loading when you switch away.
+    /// </summary>
+    /// <typeparam name="TParam">Parameter type for the producer (must provide document, settings, and replicate info)</typeparam>
+    /// <typeparam name="TResult">Result type from the producer</typeparam>
+    public class ReplicateCachingReceiver<TParam, TResult> : IDisposable where TResult : class
+    {
+        private readonly Receiver<TParam, TResult> _receiver;
+        private readonly Func<TParam, SrmDocument> _getDocument;
+        private readonly Func<TParam, object> _getSettings;
+        private readonly Func<TParam, int> _getCacheKey;
+
+        private readonly ConcurrentDictionary<int, TResult> _localCache = new ConcurrentDictionary<int, TResult>();
+        private readonly Dictionary<int, CompletionListener> _pendingListeners = new Dictionary<int, CompletionListener>();
+        private SrmDocument _cachedDocument;
+        private object _cachedSettings;
+        private int _currentCacheKey = int.MinValue;
+
+        // Exception from background completion listener, to be thrown on UI thread
+        private Exception _pendingException;
+        // Track the last reported exception to ensure we only report each error once
+        private Exception _reportedException;
+
+        /// <summary>
+        /// Creates a caching wrapper around a Receiver.
+        /// </summary>
+        /// <param name="receiver">The underlying Receiver for background computation</param>
+        /// <param name="getDocument">Function to extract document from parameters</param>
+        /// <param name="getSettings">Function to extract settings from parameters (used for cache invalidation)</param>
+        /// <param name="getCacheKey">Function to extract cache key (replicate index) from parameters</param>
+        public ReplicateCachingReceiver(
+            Receiver<TParam, TResult> receiver,
+            Func<TParam, SrmDocument> getDocument,
```

Comment:
The methods "getDocument", "getSettings" and "getCacheKey" only ever get used on objects of type `TParam`.

A different way to accomplish this would be to require that TParam inherit from a base class which has properties "Document", "Settings" and "CacheKey".